prefect_kubernetes/utilities.py

```diff
@@ -180,3 +185,126 @@ def _slugify_label_value(value: str, max_length: int = 63) -> str:
     # Kubernetes to throw the validation error
 
     return slug
+
+
+class ResilientStreamWatcher:
+    """
+    A wrapper class around kuberenetes.watch.Watch that will reconnect on
+    certain exceptions.
+    """
+
+    DEFAULT_RECONNECT_EXCEPTIONS = (urllib3.exceptions.ProtocolError,)
+
+    def __init__(
+        self,
+        logger: Optional[logging.Logger] = None,
+        max_cache_size: int = 50000,
+        reconnect_exceptions: Optional[List[Type[Exception]]] = None,
+    ) -> None:
+        """
+        A utility class for managing streams of Kuberenetes API objects and logs
+
+        Attributes:
+            logger: A logger which will be used interally to log errors
+            max_cache_size: The maximum number of API objects to track in an
+                internal cache to help deduplicate results on stream reconnects
+            reconnect_exceptions: A list of exceptions that will cause the stream
+                to reconnect.
+        """
+
+        self.max_cache_size = max_cache_size
+        self.logger = logger
+        self.watch = watch.Watch()
+
+        reconnect_exceptions = (
+            reconnect_exceptions
+            if reconnect_exceptions is not None
+            else self.DEFAULT_RECONNECT_EXCEPTIONS
+        )
+        self.reconnect_exceptions = tuple(reconnect_exceptions)
+
+    def _stream(self, cache: Optional[FIFOCache], func: Callable, *args, **kwargs):
+        """
+        A private method for streaming API objects or logs from a Kubernetes
+        client function. This method will reconnect the stream on certain
+        configurable exceptions and deduplicate results on reconnects if
+        streaming API objects.
+        """
+        keep_streaming = True
+        while keep_streaming:
+            try:
+                for event in self.watch.stream(func, *args, **kwargs):
+                    # check that we want to and can track this object
+                    if (
+                        cache is not None
+                        and isinstance(event, dict)
+                        and "object" in event
+                    ):
+                        uid = event["object"].metadata.uid
+                        if uid not in cache:
+                            cache[uid] = None
```

Comment:
Why `None` and not the event? 🤔

Reply:
Oh, I see, this isn't a cache, you're using a cache to dedupe seen objects.

Reply:
A `collections.deque` will be around 10x faster and might be worth using instead since the FIFOCache is just for tracking seen IDs. With a `maxlen` and using `append` to add items, the deque will remove items from the other end.

Reply:
Part of the reason I reached for the the FIFOCache was that I wanted efficient lookups, I'm not sure that the `deque` will support that. Maybe there is a clever way to hotwire this loop so that I can use a `deque` to track ids and prevent unbounded growth?

Reply:
Summarizing an offline exchange: we want to keep efficient lookups and don't need the FIFO behavior so we opted instead to run with a custom implementation of `set` that maintains a max size. The behavior is simple enough to not require a dependency on cachetools.